pkg/serviceimport/controller_test.go

```diff
@@ -118,26 +130,153 @@ func testLifecycleNotifications() {
+			}).Should(BeTrue())
+		})
+	})
+
+	When("single service is present in 2 clusters with 2 ips in different namespace in the same cluster which is connected", func() {
+		JustBeforeEach(func() {
+			mockCs.clusterStatusMap[clusterID] = true
+		})
+		It("should return the same ip for each namespace", func() {
+			_, _ = fakeClientset.LighthouseV2alpha1().ServiceImports(serviceImport.Namespace).Create(serviceImport)
+			si := newServiceImport(namespace2, service1, serviceIP2, clusterID)
+			_, _ = fakeClientset.LighthouseV2alpha1().ServiceImports(si.Namespace).Create(si)
+
+			Eventually(func() bool {
+				first, _ := controller.serviceImports.SelectIP(namespace1, service1, mockCs.IsConnected)
+				second, _ := controller.serviceImports.SelectIP(namespace2, service1, mockCs.IsConnected)
+				third, _ := controller.serviceImports.SelectIP(namespace1, service1, mockCs.IsConnected)
+				forth, _ := controller.serviceImports.SelectIP(namespace2, service1, mockCs.IsConnected)
+				return (first == serviceIP) && (second == serviceIP2) && (third == serviceIP) && (forth == serviceIP2)
+			}).Should(BeTrue())
+		})
+	})
+
+	When("single service is present in 3 clusters with 3 ips in 3 clusters which are connected", func() {
+		JustBeforeEach(func() {
+			mockCs.clusterStatusMap[clusterID] = true
+			mockCs.clusterStatusMap[clusterID2] = true
+			mockCs.clusterStatusMap[clusterID3] = true
+		})
+		It("should return ip for each service in its turn according to rr", func() {
+			_, _ = fakeClientset.LighthouseV2alpha1().ServiceImports(serviceImport.Namespace).Create(serviceImport)
+			si1 := newServiceImport(namespace1, service1, serviceIP2, clusterID2)
+			_, _ = fakeClientset.LighthouseV2alpha1().ServiceImports(si1.Namespace).Create(si1)
+			si2 := newServiceImport(namespace1, service1, serviceIP3, clusterID3)
+			_, _ = fakeClientset.LighthouseV2alpha1().ServiceImports(si2.Namespace).Create(si2)
+
+			Eventually(func() bool {
+				first, _ := controller.serviceImports.SelectIP(namespace1, service1, mockCs.IsConnected)
+				second, _ := controller.serviceImports.SelectIP(namespace1, service1, mockCs.IsConnected)
+				third, _ := controller.serviceImports.SelectIP(namespace1, service1, mockCs.IsConnected)
+				return (first == serviceIP) && (second == serviceIP2) && (third == serviceIP3)
+			}).Should(BeTrue())
+		})
+	})
+}
+
+func verifyCachedServiceImport(controller *Controller, expected *lighthousev2a1.ServiceImport, m *MockClusterStatus) {
+	Eventually(func() string {
 		name := expected.Annotations["origin-name"]
 		namespace := expected.Annotations["origin-namespace"]
-		ipList, ok := controller.serviceImports.GetIps(namespace, name)
-		if ok {
-			return ipList
-		}
-		return nil
-	}).Should(Equal([]string{expected.Status.Clusters[0].IPs[0]}))
+		selectedIp, _ := controller.serviceImports.SelectIP(namespace, name, m.IsConnected)
+		return selectedIp
+	}).Should(Equal(expected.Status.Clusters[0].IPs[0]))
```

Comment:
The controller is only responsible for populating the Map via Put/Remove therefore it really only needs to verify those function were called properly, ie I don't think we should be testing `SelectIP` here.  I know that `GetIps` is now removed - we can either access the internal `svcMap` directly or mock Put/Remove and use a chan to capture and verify the `ServiceImport`.  I think I prefer the latter. But again we can leave it as is for now.

Reply:
@tpantelis thanks Tom,  I'll change the tests here, and separate the test files into 2 after we merge...
I'm wondering if I need to have a separate file for the `serviceInfo` object like the Map object, with its own constructor and functions... for now I added it in the map.go file